### PR TITLE
feat: build dependency "tree" from build spec files

### PR DESCRIPTION
**Pull request by kyegupov**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?

Implements parsing of manifests with tests.
A simple flat "tree" of dependencies is produced, since we are not traversing the imports in the code.

Note: more comprehensive and negative tests to be added in a separate PR.

#### Where should the reviewer start?

parser.ts code is transplanted (with refactorings) from snyk-go-plugin
simple test displays the basic use case

#### What are the relevant tickets?

https://snyksec.atlassian.net/browse/BST-484

https://snyksec.atlassian.net/browse/BST-569

### CODE REVIEW

File: lib/index.ts [hunk 1]
```diff
@@ -1,36 +1,36 @@
...
-  dependencies: {
-    [dep: string]: DepTree;
-  };
+  dependencies?: DepDict;
   depType?: 'prod' | 'dev';
   hasDevDependencies?: boolean;
   missingLockFileEntry?: boolean;
 }
 
+export type PackageManagerType = 'dep'|'govendor';
```
Comment (orsagie) [inline lib/index.ts]: let's adhere to the ones already in CLI s/dep/golangdep

File: lib/index.ts [hunk 2]
```diff
@@ -1,36 +1,36 @@
...
-  root: string,
-  manifestFilePath: string,
-  lockFilePath: string,
-  includeDev = false,
-  strict = true,
+  lockfileType: PackageManagerType,
 ): Promise<DepTree> {
-
-  throw new Error('Not implemented');
+  const config = parseConfig(lockfileType, manifestFileContents, lockFileContents);
```
Comment (orsagie) [inline lib/index.ts]: can you call this something other than `config`, just cause that is usually reserved for an actual config. `goConfig`?

File: lib/parser.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
+import * as toml from 'toml';
+
+// TODO(kyegupov): the part below will move to snyk-go-parser
```
Comment (orsagie) [inline lib/parser.ts]: not needed todo anymore (?)

File: lib/index.ts (hunk 1 above)
Comment (kyegupov) [reply lib/index.ts]: good idea.

File: lib/parser.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+interface DepManifest {
+  ignored: string[];
+}
+
+export function parseConfig(
+  packageManager: 'golangdep'|'govendor',
+  manifestContents: string,
+  lockContents: string,
+): GoProjectConfig {
+  let config = {
```
Comment (orsagie) [inline lib/parser.ts]: same here: change variable name

File: lib/index.ts [hunk 5]
```diff
@@ -1,36 +1,33 @@
...
+  const dependencies: DepDict = {};
+  for (const dep of Object.keys(config.lockedVersions)) {
+    dependencies[dep] = {
+      name: dep,
+      version: config.lockedVersions[dep].version,
+    };
+  }
+  return {
+    name: 'root',
+    version: '0.0.0',
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply lib/index.ts]: Nope :(

File: lib/index.ts [hunk 6]
```diff
@@ -1,36 +1,33 @@
...
+  const config = parseConfig(lockfileType, manifestFileContents, lockFileContents);
+  const dependencies: DepDict = {};
+  for (const dep of Object.keys(config.lockedVersions)) {
+    dependencies[dep] = {
+      name: dep,
+      version: config.lockedVersions[dep].version,
+    };
+  }
+  return {
+    name: 'root',
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply lib/index.ts]: mostly nope (there is an undocumented field in vendor.json, most likely to be missing - added to the parser anyway). In the plugin, this is extracted in the Go code (I guess by comparing the current directory to $GOPATH)

**Pull request merged**